Crash in PDFDocument::sendPDFArrayBuffer

mcatanzaro opened the pull request:
#### 5dd594fd96bb7a5b61e7af042ffe6036fc206a73
<pre>
Crash in PDFDocument::sendPDFArrayBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=256470">https://bugs.webkit.org/show_bug.cgi?id=256470</a>

Reviewed by Tim Nguyen.

When page loading is stopped, PDFDocument::finishedParsing will be
called (which makes sense, because the parsing is finished, even if not
completed). Here sendPDFArrayBuffer will be called if the content script
has been loaded, even if there is no main resource data yet. We need
guards to ensure we don&apos;t dereference nullptr when there is no main
resource data.

Additionally, I&apos;ve added an extra guard to ensure the document loader is
not nullptr. Maybe that&apos;s not possible here -- not sure -- but since the
function returns a pointer and not a reference it seems safer to check.

* Source/WebCore/html/PDFDocument.cpp:
(WebCore::PDFDocument::sendPDFArrayBuffer):

Canonical link: <a href="https://commits.webkit.org/264114@main">https://commits.webkit.org/264114@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b1a49ffa23285682d5b912f5e12f8c67995283d5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6756 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6972 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8344 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7009 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6754 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8132 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6917 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/9912 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6873 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7463 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6197 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8439 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/4899 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6117 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13929 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/6644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6197 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8964 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6684 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5478 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6070 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1591 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10247 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6444 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/PDFDocument.cpp:
```diff
@@ -207,16 +207,16 @@ void PDFDocument::postMessageToIframe(const String& name, JSC::JSObject* data)
-    if (!arrayBuffer) {
-        ASSERT_NOT_REACHED();
-        return;
+    if (auto* documentLoader = loader()) {
```
Should we consider ASSERT() instead for the loader?

mcatanzaro replied:
Possibly! I had trouble deciding because (as per the commit message) I don't know whether the loader is guaranteed to be non-null or not. I started with this conditional, then changed it to an assert, then got cold feet and changed back to a conditional before uploading. `Document::loader` definitely sometimes returns nullptr, but I'm not sure if that can happen for a PDFDocument that is already being loaded?

mcatanzaro replied:
Any preference here?

nt1m replied:
I'd prefer an ASSERT here

nt1m commented on Source/WebCore/html/PDFDocument.cpp:
```diff
@@ -207,16 +207,16 @@ void PDFDocument::postMessageToIframe(const String& name, JSC::JSObject* data)
-        return;
+    if (auto* documentLoader = loader()) {
+        if (auto mainResourceData = documentLoader->mainResourceData()) {
+            if (auto arrayBuffer = mainResourceData->tryCreateArrayBuffer()) {
```
What happens if these 2 checks fail? Do we simply show a blank page? A fallback to our usual error codepath would be nice if possible.

mcatanzaro replied:
I wasn't able to reproduce this condition to see for sure what happens, but I would indeed expect to see a blank page. That's probably actually correct, though, because the load was stopped.

I don't see other error paths in this file; what did you have in mind regarding "usual error codepath"?

mcatanzaro replied:
Well, I suppose there could be other reasons for there to be no main resource data, like a network error or something. Hopefully those are handled elsewhere? Or is that what's missing here?

mcatanzaro replied:
I think I'll leave this alone unless you have specific suggestions for improvement.

mcatanzaro committed:
Crash in PDFDocument::sendPDFArrayBuffer
https://bugs.webkit.org/show_bug.cgi?id=256470

Reviewed by Tim Nguyen.

When page loading is stopped, PDFDocument::finishedParsing will be
called (which makes sense, because the parsing is finished, even if not
completed). Here sendPDFArrayBuffer will be called if the content script
has been loaded, even if there is no main resource data yet. We need
guards to ensure we don't dereference nullptr when there is no main
resource data.

Additionally, I've added an extra guard to ensure the document loader is
not nullptr. Maybe that's not possible here -- not sure -- but since the
function returns a pointer and not a reference it seems safer to check.

* Source/WebCore/html/PDFDocument.cpp:
(WebCore::PDFDocument::sendPDFArrayBuffer):

Canonical link: https://commits.webkit.org/264114@main
